server/indexer.go

```diff
@@ -0,0 +1,225 @@
+	authante "github.com/cosmos/cosmos-sdk/x/auth/ante"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/evmos/ethermint/rpc/backend"
+	rpctypes "github.com/evmos/ethermint/rpc/types"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/libs/log"
+	tmtypes "github.com/tendermint/tendermint/types"
+	dbm "github.com/tendermint/tm-db"
+
+	ethermint "github.com/evmos/ethermint/types"
+	evmtypes "github.com/evmos/ethermint/x/evm/types"
+)
+
+const (
+	KeyPrefixTxHash  = 1
+	KeyPrefixTxIndex = 2
+
+	// TxIndexKeyLength is the length of tx-index key
+	TxIndexKeyLength = 1 + 8 + 8
+)
+
+var _ ethermint.EVMTxIndexer = &KVIndexer{}
+
+// TxHashKey returns the key for db entry: `tx hash -> tx result struct`
+func TxHashKey(hash common.Hash) []byte {
+	return append([]byte{KeyPrefixTxHash}, hash.Bytes()...)
+}
+
+// TxIndexKey returns the key for db entry: `(block number, tx index) -> tx hash`
+func TxIndexKey(blockNumber int64, txIndex int32) []byte {
+	bz1 := sdk.Uint64ToBigEndian(uint64(blockNumber))
+	bz2 := sdk.Uint64ToBigEndian(uint64(txIndex))
+	return append(append([]byte{KeyPrefixTxIndex}, bz1...), bz2...)
+}
+
+func parseBlockNumberFromKey(key []byte) (int64, error) {
+	if len(key) != TxIndexKeyLength {
+		return 0, fmt.Errorf("wrong tx index key length, expect: %d, got: %d", TxIndexKeyLength, len(key))
+	}
+
+	return int64(sdk.BigEndianToUint64(key[1:9])), nil
+}
+
+// LoadLastBlock returns the latest indexed block number, returns -1 if db is empty
+func LoadLastBlock(db dbm.DB) (int64, error) {
+	it, err := db.ReverseIterator([]byte{KeyPrefixTxIndex}, []byte{KeyPrefixTxIndex + 1})
+	if err != nil {
+		return 0, sdkerrors.Wrap(err, "LoadLastBlock")
+	}
+	defer it.Close()
+	if !it.Valid() {
+		return -1, nil
+	}
+	return parseBlockNumberFromKey(it.Key())
+}
+
+// LoadFirstBlock loads the first indexed block, returns -1 if db is empty
+func LoadFirstBlock(db dbm.DB) (int64, error) {
+	it, err := db.Iterator([]byte{KeyPrefixTxIndex}, []byte{KeyPrefixTxIndex + 1})
+	if err != nil {
+		return 0, sdkerrors.Wrap(err, "LoadFirstBlock")
+	}
+	defer it.Close()
+	if !it.Valid() {
+		return -1, nil
+	}
+	return parseBlockNumberFromKey(it.Key())
+}
+
+// KVIndexer implements a eth tx indexer on a KV db.
+type KVIndexer struct {
+	db        dbm.DB
+	logger    log.Logger
+	clientCtx client.Context
+}
+
+// NewKVIndexer creates the KVIndexer
+func NewKVIndexer(db dbm.DB, logger log.Logger, clientCtx client.Context) *KVIndexer {
+	return &KVIndexer{db, logger, clientCtx}
+}
+
+// LastIndexedBlock returns the latest indexed block number, returns -1 if db is empty
+func (kv *KVIndexer) LastIndexedBlock() (int64, error) {
+	return LoadLastBlock(kv.db)
+}
+
+// FirstIndexedBlock returns the first indexed block number, returns -1 if db is empty
+func (kv *KVIndexer) FirstIndexedBlock() (int64, error) {
+	return LoadFirstBlock(kv.db)
+}
+
+// IndexBlock index all the eth txs in a block
```

Comment:
```suggestion
// IndexBlock index all the eth txs in a block through the following steps:
//  - Iterates over all of the Txs in Block
//  - Parses eth Tx infos from cosmos-sdk events for every TxResult
//  - Iterates over all the messages of the Tx
//  - Builds and stores a indexer.TxResult based on parsed events for every message
```

Possible suggestion for extending this comment